File: packages/core/src/ratelimit/algorithms/tokenBucket.ts
import type { Policy } from '../stores/Store.js';
import type { AlgorithmOutput } from './types.js';

export interface TokenBucketState {
  readonly tokens: number;
  readonly lastRefillMs: number;
}

/**
 * Sabit hızda dolan kova. Kısa burst'lere izin verirken ortalama hızı korur —
 * gerçek trafik için varsayılan algoritma (bkz. PLAN.md §4).
 */
export function tokenBucket(
  state: TokenBucketState | undefined,
  policy: Policy,
  now: number,
  cost: number,
): AlgorithmOutput<TokenBucketState> {
  const capacity = policy.burst ?? policy.limit;
  const refillPerMs = policy.limit / policy.windowMs;
  const lastRefillMs = state?.lastRefillMs ?? now;
  const elapsed = Math.max(0, now - lastRefillMs);
  const available = Math.min(capacity, (state?.tokens ?? capacity) + elapsed * refillPerMs);

  const allowed = available >= cost;
  const tokens = allowed ? available - cost : available;

  return {
    state: { tokens, lastRefillMs: now },
    result: {
      allowed,
      limit: policy.limit,
      remaining: Math.floor(tokens),
      retryAfterMs: allowed ? 0 : Math.ceil((cost - available) / refillPerMs),
    },
  };
}

File: packages/core/src/ratelimit/lua/tokenBucket.lua
-- KEYS[1] = bucket key
-- ARGV[1] = limit, ARGV[2] = windowMs, ARGV[3] = capacity, ARGV[4] = now, ARGV[5] = cost
local limit = tonumber(ARGV[1])
local windowMs = tonumber(ARGV[2])
local capacity = tonumber(ARGV[3])
local now = tonumber(ARGV[4])
local cost = tonumber(ARGV[5])
local refillPerMs = limit / windowMs

local raw = redis.call('HMGET', KEYS[1], 'tokens', 'ts')
local tokens = tonumber(raw[1])
local lastRefillMs = tonumber(raw[2])

if tokens == nil then
  tokens = capacity
  lastRefillMs = now
end

local elapsed = math.max(0, now - lastRefillMs)
local available = math.min(capacity, tokens + elapsed * refillPerMs)

local allowed = available >= cost
local remainingTokens = available
if allowed then
  remainingTokens = available - cost
end

redis.call('HSET', KEYS[1], 'tokens', remainingTokens, 'ts', now)
redis.call('PEXPIRE', KEYS[1], math.ceil(capacity / refillPerMs) + 1000)

local retryAfter = 0
if not allowed then
  retryAfter = math.ceil((cost - available) / refillPerMs)
end

return { allowed and 1 or 0, math.floor(remainingTokens), retryAfter }
